CondTools/L1Trigger/interface/WriterProxy.h

```diff
@@ -13,13 +14,15 @@
 
 #include <string>
 
+using namespace edm;
+
 namespace l1t {
 
   /* This is class that is used to save data to DB. Saving requires that we should know types at compile time.
  * This means that I cannot create simple class that saves all records. So, I create a base class, and template
  * version of it, that will procede with saving. This approach is the same as used in DataProxy.
  */
-  class WriterProxy {
+  class WriterProxy : public EDConsumerBase {
```

Comment:
This is almost certainly incorrect fix. There is nothing in the framework that would change its behavior based on "random class" inheriting from `EDConsumerBase`.

Reply:
My understanding and please tell me if I miss smth,
the `WriterProxy` class requires here [1] to fetch data from the EventSetup where in order to be migrated correctly to `ESGetToken` one needs to run in some constructor `esConsume()`. 
(Otherwise it throws an exception if I remember correctly)
However, you are right this do not fixes the issues of the o2o but makes the writer again to be working. 

Here are some test results with and without this change
- before change [2]
- after change [3] (if you scroll a bit upwards you can see this line 
```L1TriggerKeyExtRcd@L1TriggerKeyExt PAYLOAD TOKEN 60e764cc56d62bc1541092573df9ec142b3d6c82```
which is the result of the `WriterProxy` for the call in this line [4] for `L1TriggerKeyExtRcd`.

In short, this proxy is in reality, even tho they didn't have implemented constructors because the were using the `WriterFactory`, a consumer (as it consumes from the ES) and there is benefit to inherit from the `EDConsumerBase` for having access to the `esConsume`

[1] https://github.com/cms-sw/cmssw/pull/37602/files#diff-0078a7ded0593a3ecb4912f5f77e96836c1f587a82255c935e340b870d08d651L54-R61
[2] https://cms-conddb.cern.ch/cmsDbBrowser/logs/show_O2O_log/Prep/L1TMenu/2022-04-16%2011:02:36.821735
[3] https://cms-conddb.cern.ch/cmsDbBrowser/logs/show_O2O_log/Prep/L1TMenu/2022-04-17%2013:20:30.447741
[4] https://github.com/cms-sw/cmssw/blob/master/CondTools/L1TriggerExt/plugins/L1CondDBPayloadWriterExt.cc#L98

Reply:
> My understanding and please tell me if I miss smth,
> the `WriterProxy` class requires here [1] to fetch data from the EventSetup where in order to be migrated correctly to `ESGetToken` one needs to run in some constructor `esConsume()`.

It is not sufficient to call `esConsumes()` in "some constructor". That registration call must be done in an EDModule constructor, or in any other code called from the EDModule constructor with the help of an `edm::ConsumesCollector` object (that can be obtained in the EDModule constructor with the `consumesCollector()` call; this is documented in https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#In_ED_module)

The `edm::EDConsumerBase` class is an implementation detail of the framework, and should not be used directly in user code. The registration of consumed EventSetup products via it works only because the framework knows which kinds of classes inherit from it and can therefore ask from them the consumed EventSetup products.

Inheriting from `edm::EDConsumerBase` and calling the `esConsumes()` in an incorrect place may seem to work, like it compiles and maybe even seems to run. The registration information is, however, unknown to the framework, and the contents of the `ESGetToken` objects is incorrect, leading to the results of the data product retrieval to be similar to a `reinterpret_cast<T const *>()` on a random memory address.

Reply:
I see, seems that even if the returning hash from the db query is correct, this has avoided the pitfall just by luck and needs to be re-implemented the correct way.
Thanks for the explanation, I will revert the commit for using the `EDConsumerBase` and will try to implement it with the `Collector`